test(e2e,playwright): certification page

shabeebk opened the pull request:
With this commit, the playwright will test the certification page by iterating through the wholse set of certifications available using the slug and checks for title and other info.

Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Ref #51705

<!-- Feel free to add any additional description of changes below this line -->

shabeebk committed:
Add e2e test for certification page

With this commit, the playwright will test the certification page by
iterating through the wholse set of certifications available using the
slug and checks for title and other info.

shabeebk commented:
camperbot cant code

shabeebk commented:
Every certificate slug will be checked against generic identical tests. This can further be enhanced to incorporate subjective checks to test against specific cases like the Microsoft logo.

<img width="992" alt="Screenshot 2023-10-09 at 13 21 04" src="https://github.com/freeCodeCamp/freeCodeCamp/assets/25004086/09d0f8e8-d0d5-4389-8636-d4d1c491bcf7">

shabeebk committed:
Merge branch 'freeCodeCamp:main' into e2e/certification

shabeebk committed:
Merge branch 'freeCodeCamp:main' into e2e/certification

raditotev commented on e2e/certification.spec.ts:
```diff
@@ -0,0 +1,64 @@
+  }
+
+  test.describe(`Certification page -> test for ${certificate}`, () => {
+    test.beforeAll(async ({ browser }) => {
+      page = await browser.newPage();
+      await page.goto(`/certification/certifieduser/${certificate}`);
+    });
```
This is considered an anti-pattern. See more [here](https://github.com/freeCodeCamp/freeCodeCamp/pull/51754#discussion_r1342543051).

shabeebk replied:
Updated and pushed.

shabeebk commented:
Thanks for those remarks and suggestions. I will push the additional commit to fix and comment.

shabeebk committed:
Add social link validation and change anti-pattern

The tests was written in a way that is considred as anti pattern in the
project. Updated the tests to suite the standard. Also, added the href
validation for social  links

shabeebk commented:
Updated and pushed new commit.

shabeebk commented:
<img width="992" alt="Screenshot 2023-10-11 at 9 34 45" src="https://github.com/freeCodeCamp/freeCodeCamp/assets/25004086/faeaa218-53bb-4a08-8227-51f6f1a9fb26">

shabeebk committed:
Fix typo and update test name

raditotev commented on e2e/certification.spec.ts:
```diff
@@ -0,0 +1,67 @@
+    await expect(donation).toBeVisible();
+
+    const donationText = page.getByTestId('donation-text');
+    await expect(donationText).toHaveText(
+      `Only you can see this message. Congratulations on earning this certification. It's no easy task. Running freeCodeCamp isn't easy either. Nor is it cheap. Help us help you and many other people around the world. Make a tax-deductible supporting donation to our charity today.`
+    );
```
It's better if you get the text from the translations: `client/i18n/locales/english/translations.json`

shabeebk replied:
Fixed

raditotev commented on e2e/certification.spec.ts:
```diff
@@ -0,0 +1,67 @@
+      'https://twitter.com/intent/tweet?text=I just earned the Responsive%20Web%20Design certification @freeCodeCamp! Check it out here: https://freecodecamp.org/certification/certifieduser/responsive-web-design'
+    );
+
+    await expect(page.getByTestId('project-links')).toBeVisible();
```
Can you extend the test here with assertions for each of the links?

shabeebk replied:
Should I add the tests for all those links?

raditotev reviewed (commented):
It will be nice if you include test case for Microsoft certificate.

shabeebk commented:
@raditotev So, do you think we should test both cases, or testing against Microsoft is good enough? I could see that Microsoft has an extra logo and signee.

raditotev commented:
I think both cases should be covered.

shabeebk committed:
Add tests for microsoft certs, and url assertion

shabeebk committed:
Add url assertion for project related links

raditotev commented on e2e/certification.spec.ts:
```diff
@@ -0,0 +1,194 @@
+    await expect(fccLogo).toBeVisible();
+    // Non Microsoft certs should not have this one
+    const microsoftLogo = header.getByTestId('cert-microsoft-logo');
+    await expect(microsoftLogo).not.toBeVisible();
```
Nice!

raditotev commented on e2e/certification.spec.ts:
```diff
@@ -0,0 +1,194 @@
+    await expect(certInfoContainer).toBeVisible();
+    const certTitle = certInfoContainer.getByTestId('certification-title');
+    await expect(certTitle).toHaveText(
+      translations.certification.title['Responsive Web Design']
```
👍

raditotev reviewed (approved):
Thank you for your contribution to the page! 👍
We are happy to accept these changes and look forward to future contributions. 🎉
